### PR TITLE
fix: csp SHA hashes accumulate when using custom script-src rules

### LINKED ISSUE #4201: Too many SHA256 hashes
### Version

[v1.4.4](https://github.com/nuxt.js/releases/tag/v1.4.4)

### Reproduction link

[https://invuekelowna.condos/](https://invuekelowna.condos/)

### Steps to reproduce

We're seeing way too many SHA256 hashes in production. The SSR HTML only contains a single `<script>` tag, but 39 hashes are cramped into the header.

```json
render: {

        csp: {
            policies: {
                'default-src': ['https:'],
                'img-src': ['https:', '*.google-analytics.com'],
                'style-src': ["'self'", "'unsafe-inline'", process.env.APP_HOST, '*.typekit.net', 'cdnjs.cloudflare.com', 'fonts.googleapis.com'],
                'script-src': ["'self'", "'unsafe-inline'", process.env.APP_HOST, 'sentry.io', '*.sentry-cdn.com', '*.googletagmanager.com', '*.google-analytics.com'],
                'connect-src': ["'self'", 'sentry.io'],
                'form-action': ["'self'"],
                'frame-ancestors': ["'none'"],
            }
        },
}
```

```
'sha256-FMbXmXuys/gu6SsP6mvofSD/BGOpDurmJrL2I1bkHnk=' 
'self' 
'sha256-N2oHfMC3Qmu84zb04K5RO3/cg5ujcoxeRJH9YQ3tR84=' 
'sha256-mPczhe8H2PQW+QQ67tTk2UG3mSI3oUpoNWyAA4J7R9I=' 
'sha256-8Wq1gd6h8hJ6bQFSqYvGEid3DW//mbZBXd+FTolGfkY=' 
'sha256-zC9V4dkoVhw7cXEVOWjQccAEdN/MKGColJUyrZ9hhYI=' 
'sha256-F6hR2zjfz5aX+cSWEg5UO7OEEiYDMsmfZd9hiZDpe/A=' 
'sha256-+UH49ubmgYoXmCCjXY+iDx/oi1ofm4PN84/StCNiN5k=' 
'sha256-rVnlc4J9zZYMzsCx0sAgJRFWLx0kJM7SSWatSOPkaNs=' 
'sha256-GQxqq5jecJSKe0+qcpNjsTbGFZxvYAl/92Trcn/Tg0U=' 
'sha256-y08hFs+lXl4BlxpLURwD1/laQ3RZzfQwjb42yDlhTxE=' 
'sha256-n0pgkV6FS7OvzsLw8peHhyYETYqpMkzjfYX51hxEhsU=' 
'sha256-42AEGjgM+jtszvxsJqscprDrYDJVkGdn9sSezIG8w3I=' 
'sha256-tF9IXCO2tpb4NkZfdrrsXX/a0wI8Bo9UsaTOeKg1WTQ=' 
'sha256-urhFjM/WE6LQn+KKN0aNebDZj6jWW54tETa45frQRaA=' 
'sha256-ocvmW81CTFQB4sKJkQosfuOqT/kCGvwKp1iQG03IpHo=' 
'sha256-UWbdmlvOc+NGTX86WU6pwva10Lu9B965iLk2/IjA2kM=' 
'sha256-Achxqoh5MCq15PzEA1fVzWzeOvrxn7YTRGZLD0OtE8w=' 
'sha256-m+b4NpzWABX1Am0yXMSjDPuZ2FH9QrrHWYiO3mowNHA=' 
'sha256-GL+U99gp0eeXG3u/lvoU5H7uRPddefmaG+AL9u9jO44=' 
'sha256-gillDa0S6pvuShQnQcC0rrdiCgDbZVe8lyJbpgd4DBc=' 
'sha256-mUXTUQtSEPLuRXVzbfKFGuCOMiBzsWvoVs2DjdfzuzU=' 
'sha256-tDGzw5pdh7aHHQew+EPX0RIaH4GrOliHeYghhxKO45w=' 
'sha256-mBMX1DuaAQopWA4SEcAa/SoxFFWL/NLcquybR81vT+E=' 
'sha256-Ve7qTe6nhO17OYtBExhLAU4Qc1NlZ9/73MoYcIWaBKs=' 
'sha256-7PBWy2YgwMsUv1s72/J5tZRKxZO4cgAhqSIyZ9W5GHs=' 
'sha256-4zjq9WUhFqNFnaCqPYl7NyLE+j8jC8I+2ZHhmCKbl3w=' 
'sha256-/LO3p8jWkvBIvSVlUm3lU1GO1lQb8KHGbsxx19PynfY=' 
'sha256-YMquJrfPfZDRU8+o8ecbUvZKiLny+k8ueV34JFonYJU=' 
'sha256-xiQurpqvKUkLOPFFkj5kCJyLSTnQSqyGYsmV1arOU+I=' 
'sha256-HKhAmWlKfNO1NKb3LpOX+bXdnp93qsQCRchIkWWmxCE=' 
'sha256-q5wDnHIOul9me6e0FoQBiav+jUROu4lMKG7VMMx2EpA=' 
'sha256-W9hv/ymN6ai5wqMlhtxc2wBuWdDjxFq7TR8Or/XTVds=' 
'sha256-0l2A5a5LZ+wfP7f/uKVOgb0GCsaGsoynvZyn44s5xs4=' 
'sha256-e9gR+9jEvq38HubqKsE68a1fa+H5OIWy149yMdRB1nU=' 
'sha256-FIv6NelZC+h9yoAbp/MdVATUCwSiOOs3x84s58xyhLk=' 
'sha256-ey5hovXvLaVvp5L27Ag3aIG43U/u4XD+CJI7Mhjs8OE=' 
'sha256-4tEHlaKCZIGU35WmT0vU5hDwhOwkzCZQOR+OXHSJHB0=' 
'sha256-zJoOHpt/kJHrLMUGr+Ux4NCpjJASAp9ruDjJw/aRBOE=' 
'sha256-ufKHAvqhuJ97/WSyzz+QoFizXD4iJ/dKKKWoMxyCbek='
```

### What is expected ?

A single SHA256 hash.

### What is actually happening?

39 hashes.

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8042">#c8042</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by williamchong**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Fixes https://github.com/nuxt/nuxt.js/issues/4201

Currently when user defined a `script-src` list for CSP, the `transformPolicyObject()` would modify the CSP config object for EVERY unique SHA hash generated by nuxt. This will bloat up the CSP header until it exceed limit.

This patch fixes it by returning a new object when transforming the csp policy object.

Also the original use of `Set` seems unnecessary, not sure.

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by williamchong007** [comment]
@nnkken

**Review by pi0 — COMMENTED** [review]
LGTM.

I think the `new Set().forEach` can be replaced with a `for of` loop as we should not have duplicate assets anymore.

General Note: Using `lodash/uniq` is faster than `new Set` (But `Set` is more stable when used as a cache with `has` checks)

**Comment by williamchong007** [comment]
`cspScriptSrcHashSet` comes from https://github.com/nuxt/nuxt.js/blob/3cb9b882c0585a1ecbccfd366895c13ad9e3b0c3/packages/vue-renderer/src/renderer.js#L373 , in which it seems `Set` is unnecessary, since it can only have at most 1 entry from the if block.

Would it be simpler for everything to be in `Array` and `.concat()`?

**Comment by manniL** [comment]
I chose a set back then because we had a lot of duplicate CSP hash entries (which broke the header fairly quick, see https://github.com/nuxt/nuxt.js/pull/3574)

If that's not the case anymore, feel free to change it.

**Comment by williamchong007** [comment]
Updated with array

File: packages/server/src/middleware/nuxt.js [hunk 1]
```diff
@@ -125,30 +125,25 @@ const getCspString = ({ cspScriptSrcHashSet, allowedSources, policies, isDev })
...
 
-const transformPolicyObject = (policies, cspScriptSrcHashSet) => {
+const transformPolicyObject = (policies, cspScriptSrcHashes) => {
   const userHasDefinedScriptSrc = policies['script-src'] && Array.isArray(policies['script-src'])
 
   // Self is always needed for inline-scripts, so add it, no matter if the user specified script-src himself.
 
-  const hashAndPolicySet = cspScriptSrcHashSet
-  hashAndPolicySet.add(`'self'`)
+  let hashAndPolicyList = cspScriptSrcHashes
```
Comment (manniL) [inline packages/server/src/middleware/nuxt.js]: Do we need the re-assignment here?

Comment (williamchong007) [reply packages/server/src/middleware/nuxt.js]: rewrote this part a bit to avoid modifying input param, please look at it again

**Review by TheAlexLichter — APPROVED** [review]
Thanks a lot!

**Review by pi0 — APPROVED** [review]
LGTM

**Comment by BboyKeen** [comment]
Thanks a lot for that ! 
I was going crazy on production as it exploded the headers size and made Cloudflare reject the requests.

**Pull request merged**